### ISSUE TITLE
bad process splitting

### ISSUE THREAD

**Issue by stefco** [stefco/geco_data#25]
use named pipes instead.

``` bash
pg0:~ rdm1$ tmp_dir=$(mktemp -d)
pg0:~ rdm1$ te="min max mean"
pg0:~ rdm1$ se
security                          server_send_stats                 serverrake
securityd                         serveradmin                       serverrdoc
sed                               serverctl                         serverri
seeksize.d                        serverdiagnose                    serverruby
segedit                           serverdiagnose_get-mobility-info  servertool
select                            servererb                         serviceproxyctl
sendmail                          servergem                         set
seq                               serverinfo                        set_user_mail_opts
serialver                         serverirb                         setkey
server                            serverlog                         setquota
server-apachectl                  serverloggather                   setregion
server_mail_sender                serverrails                       setuids.d
pg0:~ rdm1$ set
set                 setkey              setregion           
set_user_mail_opts  setquota            setuids.d           
pg0:~ rdm1$ man set
pg0:~ rdm1$ help set
set: set [--abefhkmnptuvxBCHP] [-o option] [arg ...]
        -a  Mark variables which are modified or created for export.
        -b  Notify of job termination immediately.
        -e  Exit immediately if a command exits with a non-zero status.
        -f  Disable file name generation (globbing).
        -h  Remember the location of commands as they are looked up.
        -k  All assignment arguments are placed in the environment for a
            command, not just those that precede the command name.
        -m  Job control is enabled.
        -n  Read commands but do not execute them.
        -o option-name
            Set the variable corresponding to option-name:
                allexport    same as -a
                braceexpand  same as -B
                emacs        use an emacs-style line editing interface
                errexit      same as -e
                errtrace     same as -E
                functrace    same as -T
                hashall      same as -h
                histexpand   same as -H
                history      enable command history
                ignoreeof    the shell will not exit upon reading EOF
                interactive-comments
                             allow comments to appear in interactive commands
                keyword      same as -k
                monitor      same as -m
                noclobber    same as -C
                noexec       same as -n
                noglob       same as -f
                nolog        currently accepted but ignored
                notify       same as -b
                nounset      same as -u
                onecmd       same as -t
                physical     same as -P
                pipefail     the return value of a pipeline is the status of
                             the last command to exit with a non-zero status,
                             or zero if no command exited with a non-zero status
                posix        change the behavior of bash where the default
                             operation differs from the 1003.2 standard to
                             match the standard
                privileged   same as -p
                verbose      same as -v
                vi           use a vi-style line editing interface
                xtrace       same as -x
        -p  Turned on whenever the real and effective user ids do not match.
            Disables processing of the $ENV file and importing of shell
            functions.  Turning this option off causes the effective uid and
            gid to be set to the real uid and gid.
        -t  Exit after reading and executing one command.
        -u  Treat unset variables as an error when substituting.
        -v  Print shell input lines as they are read.
        -x  Print commands and their arguments as they are executed.
        -B  the shell will perform brace expansion
        -C  If set, disallow existing regular files to be overwritten
            by redirection of output.
        -E  If set, the ERR trap is inherited by shell functions.
        -H  Enable ! style history substitution.  This flag is on
            by default when the shell is interactive.
        -P  If set, do not follow symbolic links when executing commands
            such as cd which change the current directory.
        -T  If set, the DEBUG trap is inherited by shell functions.
        -   Assign any remaining arguments to the positional parameters.
            The -x and -v options are turned off.

    Using + rather than - causes these flags to be turned off.  The
    flags can also be used upon invocation of the shell.  The current
    set of flags may be found in $-.  The remaining n ARGs are positional
    parameters and are assigned, in order, to $1, $2, .. $n.  If no
    ARGs are given, all shell variables are printed.
pg0:~ rdm1$ set x 4
pg0:~ rdm1$ echo $x

pg0:~ rdm1$ for t in $te; do echo $t; done
min
max
mean
pg0:~ rdm1$ declare "x=5"
pg0:~ rdm1$ echo $x
5
pg0:~ rdm1$ for t in $te; do mkfifo "${tmp_dir}/${t}"; done
pg0:~ rdm1$ for t in $te; do echo <"${tmp_dir}/${t}" & declare "pid_${t}=$!"; done
-bash: !": event not found
pg0:~ rdm1$ for t in $te; do echo <"${tmp_dir}/${t}" & p=$!; declare "pid_${t}=${p}"; done
-bash: !: event not found
pg0:~ rdm1$ for t in $te; do cat <"${tmp_dir}/${t}" & p=$!; declare "pid_${t}=${p}"; done
-bash: !: event not found
pg0:~ rdm1$ for t in $te; do sleep 10 <"${tmp_dir}/${t}" & p=$!; declare "pid_${t}=${p}"; done
-bash: !: event not found
pg0:~ rdm1$ sleep 10 & p=$!
[1] 55951
pg0:~ rdm1$ echo $tmp_dir
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ
[1]+  Done                    sleep 10
pg0:~ rdm1$ for t in $te; do echo $tmp_dir/$t; done
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/min
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/max
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/mean
pg0:~ rdm1$ ps
  PID TTY           TIME CMD
  663 ttys000    0:00.02 -bash
35153 ttys000    0:03.29 ssh stc@pg1.phys.columbia.edu
 1299 ttys001    0:00.02 -bash
31180 ttys001    0:00.74 ssh stefan.countryman@ssh.ligo.org
55291 ttys002    0:00.09 -bash
pg0:~ rdm1$ t=max
pg0:~ rdm1$ sleep 10 <"${tmp_dir}/${t}" & p=$!; declare "pid_${t}=${p}"
-bash: !: event not found
pg0:~ rdm1$ sleep 10 <"${tmp_dir}/${t}" & p=$!
[1] 55957
pg0:~ rdm1$ sleep 10 <"${tmp_dir}/${t}" & p=$! ; declare "pid_${t}=${p}"
[2] 55958
pg0:~ rdm1$ for t in $te; do cat <"${tmp_dir}/${t}" & p=$! ; declare "pid_${t}=${p}"; done
[3] 55962
[4] 55963
[5] 55964
pg0:~ rdm1$ te
-bash: te: command not found
pg0:~ rdm1$ echo te
te
pg0:~ rdm1$ echo $te
min max mean
pg0:~ rdm1$ echo ${te// /${tmp_dir}}
min/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmax/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmean
pg0:~ rdm1$ echo ${te// / ${tmp_dir}}
min /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmax /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmean
pg0:~ rdm1$ cte=,te
pg0:~ rdm1$ cte=,"$te"
pg0:~ rdm1$ echo ${cte// / ${tmp_dir}}
,min /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmax /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmean
pg0:~ rdm1$ cte=" $te"
pg0:~ rdm1$ echo ${cte// / ${tmp_dir}}
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmin /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmax /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZmean
pg0:~ rdm1$ echo ${cte// / ${tmp_dir}/}
/var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/min /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/max /var/folders/3h/x6vswx213wd4h2ptcp9_6t_00000gs/T/tmp.NEqBYEzZ/mean
pg0:~ rdm1$ echo some stuff | tee ${cte// / ${tmp_dir}/}
some stuff
pg0:~ rdm1$ some stuff
some stuff
some stuff

[3]   Done                    cat < "${tmp_dir}/${t}"
[4]-  Done                    cat < "${tmp_dir}/${t}"
[5]+  Done                    cat < "${tmp_dir}/${t}"
pg0:~ rdm1$ rm -rf "$tmp_dir" 
[1]-  Done                    sleep 10 < "${tmp_dir}/${t}"
[2]+  Done                    sleep 10 < "${tmp_dir}/${t}"
pg0:~ rdm1$ echo ${cte// / pid_}
pid_min pid_max pid_mean
```

**Comment by stefco** [comment]
example solution from [stackoverflow](http://unix.stackexchange.com/questions/28503/how-can-i-send-stdout-to-multiple-commands):

``` bash
tmp_dir=$(mktemp -d)
mkfifo "$tmp_dir/f1" "$tmp_dir/f2"
command1 <"$tmp_dir/f1" & pid1=$!
command2 <"$tmp_dir/f2" & pid2=$!
tee "$tmp_dir/f1" "$tmp_dir/f2" | command3
rm -rf "$tmp_dir"
wait $pid1 $pid2
```

**Comment by stefco** [comment]
my attempt:

``` bash
trend_extensions="min,max,mean"

local channel_name="${1}"
local trend_extensions="${2}"
# check: are we using -f flag to force overwrites?
local _f="$("${force}" && echo '-f')"
# make a folder for the named PIPES
tmp_dir="$(mktemp -d)"
# add comma to the front of the list to make name substitutions easier
comma_trend_extensions=,"${trend_extensions}"
for te in ${trend_extensions//,/ }; do
    mkfifo "${tmp_dir}/${te}"
    geco_trend_dump -t "${te}" -c "${channel_name}" ${_f} <"${tmp_dir}/${te}" & p=$!
    declare "pid_${te}=${p}"
done
# write STDIN to each named PIPE
tee ${comma_trend_extensions// / ${tmp_dir}/}
# wait for subprocesses to complete
pids=$(for pid in ${comma_trend_extensions//,/ pid_}; do printf "${!pid} "; done)
wait ${pids}
# delete PIPES     
rm -rf "${tmp_dir}"
```

note: should probably put the force flag in the interface of the function so that there is no murky global variable stuff being passed around.

**Comment by stefco** [comment]
put into a function (to test):

``` bash
force=true

foo () {
    local channel_name="${1}"
    local trend_extensions="${2}"
    # check: are we using -f flag to force overwrites?
    local _f="$("${force}" && echo '-f')"
    # make a folder for the named PIPES
    tmp_dir="$(mktemp -d)"
    # add comma to the front of the list to make name substitutions easier
    comma_trend_extensions=,"${trend_extensions}"
    for te in ${trend_extensions//,/ }; do
        mkfifo "${tmp_dir}/${te}"
        # geco_trend_dump -t "${te}" -c "${channel_name}" ${_f} <"${tmp_dir}/${te}" & p=$!
        echo -t "${te}" -c "${channel_name}" ${_f}
        cat <"${tmp_dir}/${te}" & p=$!
        declare "pid_${te}=${p}"
    done
    # write STDIN to each named PIPE
    tee ${comma_trend_extensions//,/ ${tmp_dir}/} >/dev/null
    # delete PIPES     
    rm -rf "${tmp_dir}"
    # wait for subprocesses to complete
    pids=$(
        for pid in ${comma_trend_extensions// / pid_}; do
            printf "${!pid} "
        done
    )
    wait ${pids}
}
```

test with

``` bash
printf 'some\nexample\ntext\n' | foo "H1:SYS" min,max,mean
```

that works:

```
-t min -c H1:SYS -f
-t max -c H1:SYS -f
-t mean -c H1:SYS -f
some
example
text
some
example
text
some
example
text
```

**Comment by stefco** [comment]
``` bash
fork_process_for_each_trend_extension () {
    local channel_name="${1}"
    local trend_extensions="${2}"
    # check: are we using -f flag to force overwrites?
    local _f="$("${force}" && echo '-f')"
    # make a folder for the named PIPES
    tmp_dir="$(mktemp -d)"
    # add comma to the front of the list to make name substitutions easier
    comma_trend_extensions=,"${trend_extensions}"
    for te in ${trend_extensions//,/ }; do
        mkfifo "${tmp_dir}/${te}"
        geco_trend_dump \
                -t "${te}" \
                -c "${channel_name}" \
                ${_f} <"${tmp_dir}/${te}" & p=$!
        declare "pid_${te}=${p}"
    done
    # write STDIN to each named PIPE
    tee ${comma_trend_extensions//,/ ${tmp_dir}/} >/dev/null
    # delete PIPES     
    rm -rf "${tmp_dir}"
    # wait for subprocesses to complete
    pids=$(
        for pid in ${comma_trend_extensions// / pid_}; do
            printf "${!pid} "
        done
    )
    wait ${pids}
}
```

**Comment by stefco** [comment]
Fixed in 2d2116b
